dev-tools/ts-language-server/src/analysis/typescript-semantic/scope.ts

```diff
@@ -0,0 +1,177 @@
+import { Parser, TypeScript } from "../../parser.ts";
+
+type VariableDeclarationKeyword = "let" | "const" | "var" | "import";
+
+type Variable = {
+  keyword: VariableDeclarationKeyword;
+  node: Parser.SyntaxNode;
+  name: string;
+  definition: Parser.SyntaxNode;
+  scope: Scope;
+};
+
+const blockQuery = TypeScript.query(`
+(statement_block) @block
+`);
+
+export class Scope {
+  children: Scope[] = [];
+  variables: Variable[] = [];
+
+  constructor(
+    public node: Parser.SyntaxNode,
+    public parent: Scope | null,
+    private scopeManager: ScopeManager,
+  ) {
+    // 1. find child scopes
+    // 2. find variables
+
+    const subBlocks: Parser.SyntaxNode[] = [];
+
+    for (const child of node.namedChildren) {
+      const blockMatches = blockQuery.matches(child);
+      for (const match of blockMatches) {
+        const block = match.captures[0].node;
+        if (subBlocks.length === 0) {
+          subBlocks.push(block);
+        } else {
+          const lastBlock = subBlocks[subBlocks.length - 1];
+          // TODO: check if block is a child of lastBlock; with the hypothesis that tree-sitter does a DFS
+          if (block.endPosition.row > lastBlock.endPosition.row) {
+            // block is not a child of lastBlock
+            subBlocks.push(block);
+          } else {
+            console.warn("skipped nested subblock", block);
+          }
+        }
+      }
+
+      switch (child.type) {
+        case "variable_declaration": {
+          throw new Error("todo");
```

Comment:
question: does this and similar other statements crash the server? If so, might be best to convert to a log statement or something.

Reply:
Yes. You're right. This would crash the server.